scipy/sparse/dia.py

```diff
@@ -237,6 +237,40 @@ def sum(self, axis=None, dtype=None, out=None):
 
     sum.__doc__ = spmatrix.sum.__doc__
 
+    def _add_sparse(self, other):
+        """If other is also of type dia_matrix,
+        add all the diagonals, and return a result of type dia_matrix.
+        Otherwise, convert to csr_matrix, which is the default behaviour
+        of the _add_sparse function of parent class spmatrix.
+        """
+
+        # Check if other is also of type dia_matrix
+        if not isinstance(other, type(self)):
+            # If other is not of type dia_matrix, default to
+            # converting to csr_matrix, as is done in the _add_sparse
+            # method of parent class spmatrix
+            return self.tocsr()._add_sparse(other)
+        else:
+            # The task is to compute m = self + other
+            # Start by making an empty matrix, of the datatype
+            # that should result from adding self and other
+            m = dia_matrix(self.shape, dtype = np.result_type(self, other))
+            # Add all the stored diagonals of self
+            for d in self.offsets:
+                m.setdiag(self.diagonal(d), d)
```

Comment:
We can save on this effort by initializing `m` to be a copy of `self` with the necessary dtype:

```python
dtype = np.promote_types(self.dtype, other.dtype)
m = self.astype(dtype, copy=True)
```

Reply:
ok.

But what is the difference between ```promote_type``` and ```result_type```?

Reply:
`np.result_type` is a pretty generic function that handles a lot of cases, and it calls `np.promote_types` for the actual dtype comparison logic. Since we know that we have two sparse matrices here, it's simpler to call the more specific function directly.